File: kotlin/src/test/kotlin/com/bhengubv/circleai/tourism/TourismTest.kt
// TourismTest.kt — verifies the CircleAI.Tourism port against the C# reference.

package com.bhengubv.circleai.tourism

import com.bhengubv.circleai.companion.support.FakeCompanionSession
import kotlinx.coroutines.test.runTest
import org.junit.jupiter.api.Test
import java.math.BigDecimal
import java.time.Duration
import java.time.Instant
import kotlin.test.assertEquals
import kotlin.test.assertFailsWith
import kotlin.test.assertTrue

class TourismTest {

    @Test
    fun `attractions itineraries and bookings`() {
        val b = InMemoryTourismBoard()
        b.add(Attraction("a2", "Table Mountain", "Cape Town", "ZA", -33.9, 18.4, listOf("nature", "hike")))
        b.add(Attraction("a1", "Aquarium", "Cape Town", "ZA", -33.9, 18.4, listOf("family")))
        b.add(Attraction("a3", "Union Bldgs", "Pretoria", "ZA", -25.7, 28.2, listOf("history")))
        assertEquals(listOf("Aquarium", "Table Mountain"), b.attractionsInCity("cape town").map { it.name }) // ASC
        assertEquals(listOf("a2"), b.byTag("HIKE").map { it.attractionId }) // case-insensitive
        assertFailsWith<IllegalArgumentException> { b.attractionsInCity(" ") }
        assertFailsWith<IllegalArgumentException> { b.byTag("") }

        b.plan(Itinerary("it1", "CT 3 days", listOf(ItineraryItem(0, Duration.ofHours(9), Duration.ofHours(12), "a1", "morning"))))
        assertEquals("CT 3 days", b.getItinerary("it1")!!.title)

        b.book(TourismBooking("bk1", "it1", Instant.parse("2026-05-01T00:00:00Z"), 2, BigDecimal("5000"), "ZAR"))
        assertEquals(listOf("bk1"), b.bookings.map { it.bookingId })
    }

    @Test
    fun `domain context and adapter`() = runTest {
        assertTrue(TourismDomainContext.SYSTEM_PROMPT_SNIPPET.startsWith("[DOMAIN: Tourism]"))
        assertTrue("SATSA" in TourismDomainContext.complianceFlags)

        val fake = FakeCompanionSession()
        val a = TourismCompanionAdapter(fake)
        a.sendAsync("hi")
        assertTrue(fake.lastMessage!!.startsWith("[DOMAIN: Tourism]"))
        a.estimateBudgetAsync("Cape Town", 2, 5, "mid-range")
        assertTrue(fake.lastMessage!!.contains("Estimate budget for 2 pax, 5 days in Cape Town, mid-range standard"))
    }
}

File: rust/tests/tourism_test.rs
//! tourism_test.rs
//!
//! Ports the behaviour of `CircleAI.Tourism`: attractions by city/tag (ordered
//! by name, case-insensitive) + itineraries + bookings.

use chrono::{Duration, TimeZone, Utc};
use circle_ai::tourism::{
    Attraction, ITourismBoard, InMemoryTourismBoard, Itinerary, ItineraryItem, TourismBooking,
};

#[test]
fn attractions_in_city_ordered_by_name() {
    let board = InMemoryTourismBoard::new();
    board.add(Attraction::new("a1", "Table Mountain", "Cape Town", "ZA", -33.9, 18.4, vec!["nature".into()]));
    board.add(Attraction::new("a2", "Bo-Kaap", "cape town", "ZA", -33.9, 18.4, vec!["culture".into()]));
    board.add(Attraction::new("a3", "Union Buildings", "Pretoria", "ZA", -25.7, 28.2, vec!["history".into()]));

    let ct = board.attractions_in_city("Cape Town");
    assert_eq!(ct.len(), 2);
    assert_eq!(ct[0].name, "Bo-Kaap"); // alphabetical
    assert_eq!(ct[1].name, "Table Mountain");
}

#[test]
fn by_tag_case_insensitive() {
    let board = InMemoryTourismBoard::new();
    board.add(Attraction::new("a1", "Kirstenbosch", "Cape Town", "ZA", 0.0, 0.0, vec!["Nature".into(), "gardens".into()]));
    board.add(Attraction::new("a2", "V&A Waterfront", "Cape Town", "ZA", 0.0, 0.0, vec!["shopping".into()]));

    let nature = board.by_tag("nature");
    assert_eq!(nature.len(), 1);
    assert_eq!(nature[0].attraction_id, "a1");
}

#[test]
#[should_panic(expected = "city required")]
fn blank_city_panics() {
    InMemoryTourismBoard::new().attractions_in_city("");
}

#[test]
fn itineraries_and_bookings() {
    let board = InMemoryTourismBoard::new();
    let items = vec![ItineraryItem::new(0, Duration::hours(9), Duration::hours(12), "a1", Some("morning".into()))];
    board.plan(Itinerary::new("i1", "Weekend", items));
    assert_eq!(board.get_itinerary("i1").unwrap().items.len(), 1);

    let start = Utc.with_ymd_and_hms(2026, 5, 1, 0, 0, 0).unwrap();
    board.book(TourismBooking::new("b1", "i1", start, 2, 1999.0, "ZAR"));
    let bookings = board.bookings();
    assert_eq!(bookings.len(), 1);
    assert!((bookings[0].total_price - 1999.0).abs() < 1e-9);
}
